third_party/amd/lib/TritonAMDGPUTransforms/ReorderInstructions.cpp

```diff
@@ -17,18 +16,32 @@ namespace ttg = mlir::triton::gpu;
 // Utility functions
 //===----------------------------------------------------------------------===//
 
-// Return true if the given moduleOp contains a pure matmul problem; i.e.,
-// single dot in the main loop.
-static bool isPureMatmulProblem(triton::FuncOp funcOp) {
-  bool isMatmul = true;
-  bool foundLoop = false;
-  funcOp.walk([&](scf::ForOp forOp) -> void {
-    int counter = 0;
-    forOp.walk([&counter](triton::DotOp dotOp) { ++counter; });
-    isMatmul = (isMatmul && (counter == 1));
-    foundLoop = true;
+static SmallVector<scf::ForOp> getLeafForOps(triton::FuncOp funcOp) {
+  SmallVector<scf::ForOp> allOps;
+  funcOp->walk([&](scf::ForOp forOp) { allOps.push_back(forOp); });
+
+  SmallVector<scf::ForOp> leafOps;
+  for (scf::ForOp forOp : allOps) {
+    auto searchResult = forOp.getBody()->walk(
+        [](scf::ForOp) { return WalkResult::interrupt(); });
+    if (!searchResult.wasInterrupted())
+      leafOps.push_back(forOp);
+  }
+  return leafOps;
+}
+
+// Return true if the given ForOp contains a pure matmul problem; i.e.,
+// single dot and at least 2 glboal loads in the main loop.
+static bool isPureMatmulProblem(scf::ForOp forOp) {
+  int dotCounter = 0;
+  int loadCounter = 0;
+  forOp.walk([&](Operation *op) {
+    if (isa<triton::DotOp>(op))
+      ++dotCounter;
+    else if (isa<triton::LoadOp>(op))
+      ++loadCounter;
   });
-  return foundLoop && isMatmul;
+  return dotCounter == 1 && loadCounter >= 2;
 }
```

Comment:
[My PR](https://github.com/triton-lang/triton/pull/5346/files) had the following implementation: 

```cpp

static bool isPureMatmulProblem(scf::ForOp forOp) {
  int counter = 0;
  forOp.walk([&counter](triton::DotOp dotOp) { ++counter; });
  return counter == 1;
}
```

As you can see there was not restriction imposed on the number of loads. Why do we need it?

Reply:
Some of the optimizations this check is enabling, like `sinkSecondLoad`, do not apply if we have < 2 loads. One can argue that if we just have one load, also the optimization approaches is different. In general this function right now is applying to a lot of cases without much guardrail. This restricts its scope to only consider more canonical matmul problem so that we can gradually transit to more simple dedicated optimizations.